File: Word_Search_II.java
import java.util.*;

public class Solution {
    public List<String> findWords(char[][] board, String[] words) {
        Set<String> list = new HashSet<String>();
        for(int i=0; i<words.length; i++) {
            boolean flag = true;
            for(int x = 0; x < board.length && flag; x++){
                for(int y = 0; y<board[0].length && flag; y++){
                    boolean [][] visited = new boolean[board.length][board[0].length];
                    visited[x][y] = true;
                    if(finding(x, y, visited, 0, board, words[i])) {
                        flag = false;
                    }
                }
            }
            if(!flag) list.add(words[i]);
        }
        return new ArrayList<String>(list);
    }
    
    private final int[] dx = {0, 1, 0 , -1};
    private final int[] dy = {1, 0, -1 , 0};
    
    private boolean finding(int x, int y, boolean[][] visited, int k, char[][] board, String word) {
        if(k == word.length()) return true;
        if(word.charAt(k) == board[x][y]) {
            for(int i=0; i<dx.length; i++) {
                int xx = x + dx[i];
                int yy = y + dy[i];
                if(inBoundary(xx, yy, board.length, board[0].length) && !visited[xx][yy]) {
                    visited[xx][yy] = true;
                    if(finding(xx, yy, visited, k+1, board, word)) return true;
                    visited[xx][yy] = false;
                }
            }
        }
        return false;
    }
    
    private boolean inBoundary(int x, int y, int width, int height) {
        return x>=0 && x<width && y>=0 && y<height;
        
    }
    
    public static void main(String[] args) {
        Solution a = new Solution();
        char[][] board = {
            {'o','a','a','n'},
            {'e','t','a','e'},
            {'i','h','k','r'},
            {'i','f','l','v'}
        };
        String[] dict = {"oath","pea","eat","rain"};
        System.out.println(a.findWords(board, dict));
    }
}

import java.util.*;

class TrieNode {
    public boolean isWord;
    public TrieNode[] children;
    public TrieNode() {
        isWord = false;
        children = new TrieNode[26];
    }
}

class Trie {
    public TrieNode root;
    
    public Trie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode node = root;
        for(int i=0; i<word.length(); i++) {
            int c = word.charAt(i) - 'a';
            if (node.children[c] == null) node.children[c] = new TrieNode();
            node = node.children[c];
        }
        node.isWord = true;
    }
    
    public boolean search(String word){
        TrieNode node = root;
        for(int i=0; i<word.length(); i++) {
            int c = word.charAt(i) - 'a';
            if(node.children[c] == null) return false;
            node = node.children[c];
        }
        return node.isWord;
    }
    
    public boolean startsWith(String word){
        TrieNode node = root;
        for(int i=0; i<word.length(); i++) {
            int c = word.charAt(i) - 'a';
            if(node.children[c] == null) return false;
            node = node.children[c];
        }
        return true;
    }
}

public class Solution {
    
    private Set<String> list;
    public List<String> findWords(char[][] board, String[] words) {
        Trie trietree = new Trie();
        for(int i=0; i<words.length; i++) trietree.insert(words[i]);
        
        list = new HashSet<String>();
        
        for(int x = 0; x < board.length; x++){
            for(int y = 0; y<board[0].length; y++){
                finding(x, y, new boolean[board.length][board[0].length], board, "", trietree);
            }
        }
        
        return new ArrayList<String>(list);
    }
    
    private final int[] dx = {0, 1, 0, -1};
    private final int[] dy = {1, 0, -1, 0};
    
    private void finding(int x, int y, boolean[][] visited, char[][] board, String word, Trie trietree) {
        if(!inBoundary(x, y, board.length, board[0].length) || visited[x][y]) return;
        word += board[x][y];
        if(!trietree.startsWith(word)) return;
        if(trietree.search(word)) list.add(word);
        visited[x][y] = true;
        for(int i=0; i<dx.length; i++) {
            int xx = x + dx[i];
            int yy = y + dy[i];
            finding(xx, yy, visited, board, word, trietree);
        }
        visited[x][y] = false;
    }
    
    private boolean inBoundary(int x, int y, int width, int height) {
        return x>=0 && x<width && y>=0 && y<height;
        
    }
}

File: Swift/word-search-ii.swift
/**
 * https://leetcode.com/problems/word-search-ii/
 * 
 * 
 */ 
class Solution {
    
    let dir:[(Int, Int)] = [(1,0), (0,1), (-1,0), (0,-1)]
    
    /**
     * The reason why TrieNode has to be class, is the root of Trie has be an reference type object, 
     * which would be the same object shared every time being used, e.g insert, search or startWith.
     *
     * The struct will make another copy of the root instead, which doesn't server as we proposed.
     */
    class TrieNode {
        var isWord: Bool
        var children: [String: TrieNode]
        
        init() {
            self.isWord = false
            children = [String: TrieNode]()
        }
    }
    
    struct Trie {
        let root: TrieNode
        
        init() {
            self.root = TrieNode()
        }
        
        func insert(word: String) {
            var node = self.root
            let wList = word.characters.map() {"\($0)"}
            for w in wList {
                if node.children[w] == nil {
                    node.children[w] = TrieNode()
                }
                node = node.children[w]!
            }
            node.isWord = true
        }
        
        func search(word: String) -> Bool {
            var node = self.root
            let wList = word.characters.map() {"\($0)"}
            for w in wList {
                if let xNode = node.children[w] {
                    node = xNode
                } else {
                    return false
                }
            }
            return node.isWord
        }
        
        func startWith(word: String) -> Bool {
            var node = self.root
            let wList = word.characters.map() {"\($0)"}
            for w in wList {
                if let xNode = node.children[w] {
                    node = xNode
                } else {
                    return false
                }
            }
            return true
        }
        
        func printTrie() {
            printNode(self.root, "")
        }
        
        private func printNode(_ node: TrieNode, _ s: String) {
            if node.isWord {
                print(s)
            }
            
            for (w, child) in node.children {
                printNode(child, s + w)
            }
        }
    }
    
    var wSet: Set<String>!
    
    func findWords(_ board: [[Character]], _ words: [String]) -> [String] {
        wSet = Set<String>()
        let n = board.count
        if n == 0 {
            return [String]()
        }
        let m = board[0].count
        
        let node = Trie()
        for w in words {
            node.insert(word: w)
        }
        
        for x in 0..<n {
            for y in 0..<m {
                finding(x: x, y: y, Array(repeating: Array(repeating: false, count: m), count: n), board, "", node)
            }
        }
        return Array(wSet)
    }
    
    func finding(x: Int, y: Int, _ visited: [[Bool]], _ board: [[Character]], _ word: String, _ node: Trie) {
        let w = word + String(board[x][y])
        if !node.startWith(word: w) {
            return
        }
        if node.search(word: w) {
            wSet.insert(w)
        }
        var v = visited
        v[x][y] = true
        
        for (dx, dy) in dir {
            let xx = x + dx
            let yy = y + dy
            if xx>=0, xx<board.count, yy>=0, yy<board[0].count, !visited[xx][yy] {
                finding(x: xx, y: yy, v, board, w, node)
            }
        }
    }
}
let board = ["oaan","etae","ihkr","iflv"].map({Array($0.characters)})
let words = ["oath","pea","eat","rain", "an"]

print("\(findWords(board, words))")

let trie = Trie()
for w in ["abc", "acd", "acf"] {
    print("Word: \(w)")
    trie.insert(word: w)
}

trie.printTrie()
/**
 * https://leetcode.com/problems/word-search-ii/
 * 
 * 
 */ 
// Date: Tue Jun 30 11:33:53 PDT 2020
class Solution {
    class TrieNode {
        var isWord: Bool 
        var children: [String: TrieNode]
        
        convenience init() {
            self.init(isWord: false)
        }
        
        init(isWord: Bool) {
            self.isWord = isWord
            self.children = [:]
        }
    }
    
    class Trie {
        let root: TrieNode
        init() {
            self.root = TrieNode()
        }
        
        func insert(_ word: String) {
            var node = self.root
            for w in word {
                if let child = node.children[String(w)] {
                    node = child
                } else {
                    node.children[String(w)] = TrieNode()
                    node = node.children[String(w)]!
                }
            }
            node.isWord = true
        }
        
        func start(with word: String) -> Bool {
            var node = self.root
            for c in word {
                if let child = node.children[String(c)] {
                    node = child
                } else {
                    return false
                }
            }
            return true
        }
        
        func search(_ word: String) -> Bool {
            var node = self.root
            for c in word {
                if let child = node.children[String(c)] {
                    node = child
                } else {
                    return false
                }
            }
            return node.isWord
        }
    }
    
    func findWords(_ board: [[Character]], _ words: [String]) -> [String] {
        let trie = Trie()
        for word in words {
            trie.insert(word)
        }
        
        var result: Set<String> = []
        
        let n = board.count
        guard let m = board.first?.count else { return [] }
        
        func visit(_ x: Int, _ y: Int, _ prefix: String, _ visited: inout [[Bool]]) {
            let word = prefix + String(board[x][y])
            if trie.start(with: word) == false { return }
            if trie.search(word) { result.insert(word) }
            visited[x][y] = true
            let dx = [1, 0, -1, 0]
            let dy = [0, 1, 0, -1]
            
            for index in 0 ..< dx.count {
                let xx = x + dx[index]
                let yy = y + dy[index]
                if xx >= 0, xx < n, yy >= 0, yy < m, visited[xx][yy] == false {
                    visit(xx, yy, word, &visited)
                }
            }
            visited[x][y] = false
        }
        
        for x in 0 ..< n {
            for y in 0 ..< m {
                var visited = Array(repeating: Array(repeating: false, count: m), count: n)
                visit(x, y, "", &visited)
            }
        }
        
        return Array(result)
    }
}
